src/python/pants/engine/fs_test.py

```diff
@@ -1508,6 +1509,29 @@ def test_snapshot_diff(
     assert diff.changed_files == expected_diff.changed_files
 
 
+@contextmanager
+def retry_assertions(n: int, sleep_duration: float = 0.05):
+    """Retry the inner block if any assertions failed.
+
+    This is used to handle any failures resulting from an external system not fully processing
+    certain events as expected.
+    """
+    last_exception: BaseException | None = None
+
+    while n > 0:
+        try:
+            yield
+            return
+        except AssertionError as e:
+            last_exception = e
+            n -= 1
+            time.sleep(sleep_duration)
+            sleep_duration *= 2
+
+    assert last_exception is not None
+    raise last_exception
```

Comment:
Nice approach! Unfortunately I think Python doesn't make it quite this easy 😢 a `contextmanager` is only permitted to `yield` once.

I put this into a file:
```python
import time
from contextlib import contextmanager


@contextmanager
def retry_assertions(n: int, sleep_duration: float = 0.05):
    """Retry the inner block if any assertions failed.

    This is used to handle any failures resulting from an external system not fully processing
    certain events as expected.
    """
    last_exception: BaseException | None = None

    while n > 0:
        try:
            yield
            return
        except AssertionError as e:
            last_exception = e
            n -= 1
            time.sleep(sleep_duration)
            sleep_duration *= 2

    assert last_exception is not None
    raise last_exception


with retry_assertions(10):
    assert False
```

And ran it like `python retry.py`, and Python rejected it:

```
Traceback (most recent call last):
  File "/Users/huon/tmp/retry.py", line 16, in retry_assertions
    yield
  File "/Users/huon/tmp/retry.py", line 29, in <module>
    assert False
AssertionError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/huon/tmp/retry.py", line 28, in <module>
    with retry_assertions(10):
  File "/Users/huon/.pyenv/versions/3.10.4/lib/python3.10/contextlib.py", line 185, in __exit__
    raise RuntimeError("generator didn't stop after throw()")
RuntimeError: generator didn't stop after throw()
```

One way to do it is to have a iterable construct that yields context managers, e.g.: https://tenacity.readthedocs.io/en/latest/#retrying-code-block

That's pretty awkward.

Other ways might be:

- write the loop in the test directly 😢 
- turn this into a function decorator, and have a `def assert_not_found() -> None:` function that is decorated and called
- have the helper be `retry_until_not_none` or something, so it can take a lambda (by side stepping using the `assert`): `m3 = retry_until_not_none(lambda: get_metadata("not-found"), 3)`

Reply:
I went with the alternative of just passing in a `Callable[[], Any]`. Test passes locally now and the latest push should work.